single_kernel_mongo/managers/ldap.py

```diff
@@ -0,0 +1,162 @@
+        substrate: Substrates,
+        relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP,
+        cert_relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP_CERT,
+    ):
+        super().__init__(parent=dependent, key=relation_name)
+        self.dependent = dependent
+        self.charm = dependent.charm
+        self.workload = self.dependent.workload
+        self.state = state
+        self.substrate = substrate
+        self.relation_name = relation_name
+        self.cert_relation_name = cert_relation_name
+        self.ldap_requirer = LdapRequirer(self.charm, self.relation_name)
+        self.certificate_transfer = CertificateTransferRequires(self.charm, self.cert_relation_name)
+
+    def assert_pass_hook_checks(self) -> None:
+        """Runs some hook checks before allowing the hook to run."""
+        if not self.state.db_initialised:
+            raise DeferrableFailedHookChecksError("DB is not initialised")
+        if self.state.is_role(MongoDBRoles.SHARD):
+            self.charm.status_manager.to_blocked("Cannot integrate LDAP with shard.")
+            raise NonDeferrableFailedHookChecksError("Cannot integrate LDAP with shard.")
+        if self.state.upgrade_in_progress:
+            raise DeferrableFailedHookChecksError(
+                "Adding LDAP is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+        return
+
+    def is_valid_ldap_integration(self) -> bool:
+        """Returns True if the integration to ldap is valid."""
+        return (self.state.ldap_relation is not None) and not self.state.is_role(MongoDBRoles.SHARD)
+
+    def on_ldap_ready(self, relation: Relation) -> None:
+        """Runs when LDAP is ready."""
+        self.assert_pass_hook_checks()
+
+        self.charm.status_manager.to_maintenance("Configuring LDAP")
+
+        ldap_data = self.ldap_requirer.consume_ldap_relation_data(relation=relation)
+        if not ldap_data:
+            logger.info("Waiting for LDAP data.")
+            raise WaitingForLdapDataError("Waiting for LDAP data.")
+
+        if not ldap_data.ldaps_urls:
+            logger.info("You must configure Glauth with LDAPS support")
+            raise LDAPSNotEnabledError("LDAPS not enabled.")
+
+        if self.charm.unit.is_leader():
+            self.state.ldap.set_from(ldap_data)
+
+        if self.state.ldap.is_ready():
+            logger.info("Restarting mongodb server for LDAP integration")
+            self.dependent.restart_charm_services()
+
+            self.charm.status_manager.to_active()
+
+    def on_ldap_unavailable(self) -> None:
+        """Runs when the LDAP integration is broken."""
+        self.state.ldap.clean_databag()
+
+        self.dependent.restart_charm_services()
+
+    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
+        """Runs when we receive the LDAP certificates."""
+        self.assert_pass_hook_checks()
```

Comment:
don't shards need these certs? Since clients eventually communicate with shards?

Reply:
This will be done in another PR if needed.

Keep in mind that this is for identity users, not client applications.
It is not recommended to log in directly on a shard, but use mongos instead.
The decision was taken not to integrate with shards, and sharing the necessary information will be done if necessary by the config-server.

As I said, this PR introduces the manager, it does not even update the configuration of the services.
I'm trying to keep small PR sizes and minimal changes to reduce the risk of breaking things.
On top of that, since this PR does not listen to events, it is for now completely dead code which make continuous development easier, not breaking the charms.